src/clj/puppetlabs/services/legacy_routes/legacy_routes_core.clj

```diff
@@ -0,0 +1,144 @@
+(defn- map-accept-raw-to-binary
+  "Convert Accept: raw media types to Accept: binary.  NOTE: This method does
+  not conform to RFC-2616 Accept: header specification, but then again
+  neither does Accept: raw or Accept: binary
+
+  The split on `,` is naive, the method should take into account `;` parameters
+  and the such to be RFC compliant, but puppet agent does not use this format
+  so implementing the full spec is unnecessary."
+  [request]
+  (if-let [accept (get-in request [:headers "accept"])]
+    (let [vals (str/split accept #"\s*,\s*")]               ; BAD! See RFC-2616
+      (assoc-in request [:headers "accept"]
+        (str/join ", " (distinct (replace {"raw" "binary"} vals)))))
+    request))
+
+(defn- map-accept-s-pson-to-binary
+  "Convert Accept: s, pson media types to Accept: binary given a ring request
+  map.  This method should only be used to munge file_bucket_file API requests
+  from legacy agents speaking to a contemporary master."
+  [request]
+  (if-let [accept (get-in request [:headers "accept"])]
+    (let [v4-accept-val "binary"
+          vals (str/split accept #"\s*,\s*")                ; BAD! see RFC-2616
+          munged-accept (->> vals
+                          (replace {"s" v4-accept-val "pson" v4-accept-val})
+                          (distinct)
+                          (str/join ", "))]
+      (assoc-in request [:headers "accept"] munged-accept))
+    request))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Handler wrapper
+
+(defn request-compatibility-wrapper
+  "Given a Puppet 3.X CA request, convert it to a Puppet 4.X CA request."
+  [handler mount-point api-version request]
+  (let [{{environment :environment} :params
+         path-info                  :path-info
+         query-string               :query-string} request
+        query-params (if query-string (ring-codec/form-decode query-string) {})]
+    (let [compat-request
+          (-> request
+              (map-accept-raw-to-binary)
+              (assoc :path-info    (str "/" api-version path-info)
+                     :context      mount-point
+                     :uri          (str mount-point "/" api-version path-info)
+                     :query-string (ring-codec/form-encode
+                                     (merge
+                                       query-params
+                                       (when environment {:environment environment}))))
+              (dissoc :params :route-params))]
+      (handler compat-request))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Routing
+
+(defn v2_0-routes
+  "Creates the compojure routes to handle the 3.X master's '/v2.0' routes."
+  [request-handler]
+  (compojure/routes
+    (compojure/GET "/environments" request
+                   (request-handler request))))
+
+(defn legacy-master-routes
+  [master-request-handler ca-request-handler]
+  (compojure/routes
```

Comment:
RE: the header remapping, I'm wondering if enough of the routes here require some header remapping and if the mappings are always 1->1, like just raw and s -> binary, if it would make sense to just have all of these routes call through a common function that does the header translation before calling through to `master-request-handler`.  Just concerned we might be missing some subtle case for some endpoints that the custom translation we're doing per endpoint here wouldn't cover.